docs/Blog-Post.md

```diff
@@ -0,0 +1,276 @@
+* Named axes allow you to abstract over unreferenced dimensions, making code more flexible for things like multi-headed attention or flexible attention masking.
+
+Let's take a look at a practical example of how Haliax can be used.
+Here's a minimal attention module implementation in Haliax. For a more detailed introduction,
+please see the [Haliax tutorial](https://colab.research.google.com/drive/1TiTcQQ4V5mopbgCu1SVl-oqJtXn7rFnC).
+
+```python
+import equinox as eqx
+import jax
+import jax.numpy as jnp
+from jax.random import PRNGKey
+import haliax as hax
+import haliax.nn as hnn
+
+# Named Axes for Tensor Dimensions
+Pos = hax.Axis("position", 1024)  # sequence
+KPos = Pos.alias("key_position")  # key sequence for attention
+Head = hax.Axis("head", 8)  # number of attention heads
+Key = hax.Axis("key", 64)  # key size
+Embed = hax.Axis("embed", 512)  # embedding size
+
+# Despite making no reference to batching or heads, this same implementation is also batch-capable and multi-headed
+# (or multi-query) and even supports attending to or from non-sequential keys (e.g. image patches)
+def attention(Key, KPos, query, key, value, mask):
+  # how similar is each query to each key
+  scores = hax.dot(Key, query, key) / jnp.sqrt(Key.size)
+
+  # mask out invalid positions
+  if mask is not None:
+    scores -= 1E9 * (1.0 - mask)
+
+  # convert to probabilities
+  scores = hax.nn.softmax(scores, KPos)
+
+  # weighted sum of values
+  result = hax.dot(KPos, scores, value)
+
+  return result
+
+
+# Causal Mask means that if pos >= key_pos, then pos can attend to key_pos
+causal_mask = hax.arange(Pos).broadcast_axis(KPos) >= hax.arange(KPos)
+
+class Attention(eqx.Module):
+  proj_qkv: hnn.Linear  # input projection from [Embed] -> [(q, k, v), Head, Key]
+  proj_answer: hnn.Linear  # output projection from [Head, Key] -> [Embed]
+
+  @staticmethod
+  def init(Embed, Head, Key, *, key: PRNGKey):
+    Qkv = hax.Axis("qkv", 3)  # create all three at once
+
+    k_qkv, k_ans = jax.random.split(key, 2)
+    proj_qkv = hnn.Linear.init(In=Embed, Out=(Qkv, Head, Key), key=k_qkv)
+    proj_answer = hnn.Linear.init(In=(Head, Key), Out=Embed, key=k_ans)
+    return Attention(proj_qkv, proj_answer)
+
+  def __call__(self, x, mask):
+    qkv_out = self.proj_qkv(x)
+    q, k, v = qkv_out.unbind("qkv")
+
+    # Rename k and v's Pos as Haliax doesn't support unnamed axes or duplicate axes
+    k = k.rename({Pos: KPos})
+    v = v.rename({Pos: KPos})
+
+    answers = attention(Key, KPos, q, k, v, causal_mask)
+
+    x = self.proj_answer(answers)
+    return x
+```
+
+We use named axes both to improve legibility and to enable scale: named axes are the basis of our
+[Fully-Sharded Data Parallel](https://engineering.fb.com/2021/07/15/open-source/fsdp/) implementation. FSDP can
+be added to the training code with about 10 lines of code, enabling scale to at least 256 TPU cores (which is
```

Comment:
Would be cool to show those 10 lines

Reply:
it's 10 lines of code that are distributed across 5-6 places :-(, they mostly look the same though